test/Elastic.Apm.Tests.MockApmServer/MockApmServer.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using System.Diagnostics.Contracts;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Apm.Helpers;
+using Elastic.Apm.Logging;
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Elastic.Apm.Tests.MockApmServer
+{
+	public class MockApmServer
+	{
+		internal readonly ReceivedData ReceivedData = new ReceivedData();
+		private readonly string _dbgCurrentTestName;
+
+		internal MockApmServer(IApmLogger logger, string dbgCurrentTestName)
+		{
+			Logger = logger.Scoped(nameof(MockApmServer));
+			_dbgCurrentTestName = dbgCurrentTestName;
+		}
+
+		private static class PortScanRange
+		{
+			internal const int Begin = 23_456;
+			internal const int End = 24_567; // not included
+		}
+
+		private CancellationTokenSource _cancellationTokenSource;
+		private int _port;
+		private Task _runningTask;
+
+		internal IApmLogger Logger { get; }
+
+		internal int FindAvailablePortToListen()
+		{
+			var numberOfPortsTried = 0;
+			var numberOfPortsInScanRange = PortScanRange.End - PortScanRange.Begin;
+			var currentPort = RandomGenerator.GetInstance().Next(PortScanRange.Begin, PortScanRange.End);
+			while (true)
+			{
+				++numberOfPortsTried;
+				try
+				{
+					Logger.Debug()?.Log("Trying to listen on port {PortNumber}...", currentPort);
+					var listener = new HttpListener();
+					listener.Prefixes.Add($"http://localhost:{currentPort}/");
+					listener.Start();
+					listener.Stop();
+					Logger.Debug()?.Log("Port {PortNumber} is available - it will be used to accept connections from the agent", currentPort);
+					return currentPort;
+				}
+				catch (HttpListenerException ex)
+				{
+					Logger.Debug()
+						?.LogException(ex, "Failed to listen on port {PortNumber}. " +
+							"Number of ports tried so far: {NumberOfPorts} out of {NumberOfPorts}",
+							currentPort, numberOfPortsTried, numberOfPortsInScanRange);
+					if (numberOfPortsTried == numberOfPortsInScanRange)
+					{
+						throw new InvalidOperationException("Could not find an available port for Mock APM Server to listen. " +
+							$"Ports range that was tried: [{PortScanRange.Begin}, {PortScanRange.End})");
+					}
+				}
+
+				if (currentPort + 1 == PortScanRange.End) currentPort = PortScanRange.Begin;
+				else ++currentPort;
+			}
+		}
+
+		internal void RunAsync(int port)
```

Comment:
Just a really minor note: I think the convention is to only postfix real async methods with `Async`, so methods that return a `Task` or use `await`, which is not the case here. I'd leave it like this, alternatively it can be something like `RunInBackground`.

Reply:
Done.